File: src/0678-valid-parenthesis-string/solution.java
class Solution {

    public boolean checkValidString(String s) {
        int min = 0, max = 0;

        for(char c : s.toCharArray()) {
            if (c == '(') {
                min++;
                max++;
            }
            else if (c == ')') {
                min = Math.max(0, min - 1);
                max--;
                if (max < 0) return false;
            }
            else {
                min = Math.max(0, min - 1);
                max++;
            }
        }

        return min == 0;
    }
}

File: src/0678-valid-parenthesis-string/solution.py
class Solution:
    def checkValidString(self, s: str) -> bool:
        amin, amax = 0, 0

        for c in s:
            if c == "(": amin, amax = amin + 1, amax + 1
            elif c == ")": amin, amax = amin - 1, amax - 1
            else: amin, amax = amin - 1, amax + 1
            amin = max(0, amin)
            if amax < 0: return False

        return amin == 0

